File: solutions/0093_Restore_IP_Addresses/backtracking.py
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        self.temp = []
        self.ret = []
        self.recur(s, 0, 0)
        return self.ret

    def recur(self, s, i, d):
        if d >= 4:
            if i == len(s):
                self.ret.append('.'.join(self.temp))
            return
        if i+1<=len(s) and '0'<=s[i:i+1]<='9':
            self.temp.append(s[i])
            self.recur(s, i+1, d+1)
            self.temp.pop()
        if i+2<=len(s) and '10'<=s[i:i+2]<='99':
            self.temp.append(s[i:i+2])
            self.recur(s, i+2, d+1)
            self.temp.pop()
        if i+3<=len(s) and '100'<=s[i:i+3]<='255':
            self.temp.append(s[i:i+3])
            self.recur(s, i+3, d+1)
            self.temp.pop()

File: solutions/0017_Letter_Combinations_of_a_Phone_Number/backtracking.java
class Solution {

    private List<String> ret;

    public List<String> letterCombinations(String digits) {
        if (digits.isEmpty()) {
            return new ArrayList<>();
        }

        Map<Integer, String> dict = new HashMap<>();
        dict.put(2, "abc");
        dict.put(3, "def");
        dict.put(4, "ghi");
        dict.put(5, "jkl");
        dict.put(6, "mno");
        dict.put(7, "pqrs");
        dict.put(8, "tuv");
        dict.put(9, "wxyz");
        dict.put(0, " ");

        ret = new ArrayList<>();
        recur(digits, 0, dict, new Stack<>());
        return ret;
    }

    private void recur(String digits,
                       int index,
                       Map<Integer, String> dict,
                       Stack<Character> stack) {
        if (index >= digits.length()) {
            ret.add(convert(stack));
            return;
        }
        Integer num = Integer.valueOf(digits.charAt(index)) - '0';
        for (char c : dict.get(num).toCharArray()) {
            stack.push(c);
            recur(digits, index + 1, dict, stack);
            stack.pop();
        }
    }

    private String convert(Stack<Character> stack) {
        StringBuilder sb = new StringBuilder();
        for (Character c : stack) {
            sb.append(c);
        }
        return sb.toString();
    }
}
